compiler/resolution/functionResolution.cpp

```diff
@@ -11255,6 +11255,82 @@ static void checkNoVoidFields()
   }
 }
 
+/************************************* | **************************************
+*                                                                             *
+*  Find specially named methods such as 'hash' that are placed on types that  *
+*  don't have the corresponding interface. This is executed here because      *
+*  resolution cleans up unused functions and interface statements, but we want*
+*  to warn even if a function isn't used.                                     *
+*                                                                             *
+************************************** | *************************************/
+
+struct SpeciallyNamedMethodInfo {
+  std::map<std::string, FnSymbol*> speciallyNamedMethods;
+};
+
+using SpeciallyNamedMethodKey = std::pair<AggregateType*, InterfaceSymbol*>;
+using SpecialMethodMap = std::map<SpeciallyNamedMethodKey, SpeciallyNamedMethodInfo>;
+
+static void checkSpeciallyNamedMethods() {
+  static const std::unordered_map<std::string, InterfaceSymbol*> reservedNames = {
+    { "hash", gHashable },
+  };
+
+  SpecialMethodMap flagged;
+
+  // TODO: for now, this will simply not warn for classes, because all classe
+  // implement hashable, and because overriding hash should be allowed. When
+  // other special methods are converted, this logic will need to be updated
+  // to handle them.
+
+  for_alive_in_Vec(FnSymbol, fn, gFnSymbols) {
+    if (!fn->isMethod()) continue;
+    if (fn->isCompilerGenerated() || fn->hasFlag(FLAG_FIELD_ACCESSOR)) continue;
+
+    auto reservedIter = reservedNames.find(fn->name);
+    if (reservedIter == reservedNames.end()) continue;
+    auto receiverType = fn->getReceiverType();
+
+    auto at = toAggregateType(receiverType);
+    if (!at || !at->isRecord()) continue;
+
+    while (at->instantiatedFrom) at = at->instantiatedFrom;
+    auto key = SpeciallyNamedMethodKey(at, reservedIter->second);
+    flagged[key].speciallyNamedMethods[fn->name] = fn;
```

Comment:
I'm not completely understanding how this line works.  Is there a file missing that defines `.speciallyNamedMethods` on a `SpecialMethodMap` or does the file already exist and I need to go look for it?

Reply:
The _values_ of `SpecialMethodMap` define this method; the top-level data structure is actually a map of maps.

```C++
struct SpeciallyNamedMethodInfo {
  std::map<std::string, FnSymbol*> speciallyNamedMethods;
};
```

The map is of { Type, Interface } -> { method name of interface -> the function where it's defined }.

Reply:
D'oh, I see it in the type definition now.  Sorry!